### DISCUSSION TITLE
Compile C/C++ code statically and link together at build time for NativeAOT?

### DISCUSSION THREAD

**Discussion by stevefan1999-personal** [dotnet/runtime#96937] [Q&A] [answered]
For example, I want to bundle sqlite3 statically with NativeAOT rather than with dynamic libraries, so that we could enjoy link time optimization, right now this is what I do:
```
	<ItemGroup>
		<PackageReference Include="SQLitePCLRaw.bundle_sqlite3" Version="2.1.7" />
	</ItemGroup>

	<ItemGroup Condition="'$(Sqlite3Library)' == 'static'">
		<DirectPInvoke Include="sqlite3"/>
		<SdkNativeLibrary Include="F:\sqlite\sqlite3.lib" />
	</ItemGroup>
```
I have to compile the `sqlite3.lib` myself as a static library as sqlite team doesn't provider it currently (and obviously for configuration reasons), with the amalgamated source code with some compiler options as follows:
```
F:\sqlite>dir
 Volume in drive F is Dev
 Volume Serial Number is 8A30-9340

 Directory of F:\sqlite

13/01/2024  04:44 pm    <DIR>          .
13/01/2024  04:44 pm    <DIR>          ..
24/11/2023  09:33 pm         8,916,712 sqlite3.c
24/11/2023  09:33 pm           637,671 sqlite3.h
24/11/2023  09:33 pm            38,149 sqlite3ext.h
               3 File(s)      9,592,532 bytes
               2 Dir(s)  149,356,875,776 bytes free

F:\sqlite> cl /c /EHsc .\sqlite3.c /DSQLITE_ENABLE_SNAPSHOT /DSQLITE_ENABLE_COLUMN_METADATA
...
F:\sqlite> lib .\sqlite3.obj
...
F:\sqlite>dir
 Volume in drive F is Dev
 Volume Serial Number is 8A30-9340

 Directory of F:\sqlite

13/01/2024  04:44 pm    <DIR>          .
13/01/2024  04:44 pm    <DIR>          ..
24/11/2023  09:33 pm         8,916,712 sqlite3.c
24/11/2023  09:33 pm           637,671 sqlite3.h
13/01/2024  04:22 pm         3,064,466 sqlite3.lib
13/01/2024  04:22 pm         2,936,205 sqlite3.obj
24/11/2023  09:33 pm            38,149 sqlite3ext.h
               5 File(s)     15,593,203 bytes
               2 Dir(s)  149,347,602,432 bytes free
```
(Technically speaking, just supplying an object file is good enough as there is just one sqlite3.c translation unit)

But this got me thinking, why can't we also bundle the C/C++ source code as part of a MSBuild process to compile those file on the fly and add them to SdkNativeLibrary altogether? Like we can invoke an external compiler to run the MSVC to build the sqlite3 library. Kinda like how you can link libraries together with vcxproj.

So basically what I wanted is something similar to the [`cmake`](https://github.com/rust-lang/cmake-rs) crate or [`cc`](https://github.com/rust-lang/cc-rs) in dotnet

Related discussion: https://github.com/ericsink/SQLitePCL.raw/discussions/582

**Comment by Clockwork-Muse** [comment]
You _can_ run arbitrary commands with the [`Exec` task](https://learn.microsoft.com/en-us/visualstudio/msbuild/exec-task?view=vs-2022):
>```csproj
>  <Target Name="ExecTest" BeforeTargets="Build">
>    <Exec Command="echo 'Hello World'"/>
>    <Exec Command="touch /somepath/something.txt"/>
>  </Target>
>```

Obviously, this isn't as good as an actual integration.... at minimum you'd have to play around with targets and other things to get it to handle build dependencies properly.
MSBuild proper has C++ compiler tasks, but not in the `dotnet` tool.

**Comment by MichalStrehovsky** [comment, marked as answer]
For a sample of how to combine C# and C++ projects, see https://github.com/dotnet/samples/tree/main/core/interop/IDynamicInterfaceCastable/src.

> so that we could enjoy link time optimization

Link time optimizations assume that instead of native code, the inputs to the linker contain special bytecode known to the Visual C++ compiler (linker will invoke the Visual C++ compiler to generate native code). Native AOT compiler doesn't produce outputs in LTO format because the LTO format is Visual C++ compiler specific. I don't think you'll see benefits of LTO outside sqlite3.lib (which will be the only compiland with LTO bytecode).

> SdkNativeLibrary 

I suggest using [the documented](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/interop#linking) `<NativeLibrary>` element instead of this undocumented item.
